### PR TITLE
[ASA] Add week 4 content

### PR DESCRIPTION
**Pull request by randomicecube**
TODO List:

- [X] Bellman-Ford's Algorithm
- [X] Shortest Path in a DAG
- [ ] Add Teacher's Notes (whenever available)
- [ ] General cleanup, typo fixing and dtc review

### CODE REVIEW

**Comment by randomicecube** [comment]
@diogotcorreia I think this is ready for review, missing teacher's notes (they aren't available atm) which can be added whenever they're made available

File: content/asa/0005-caminhos-mais-curtos.md [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+![Sub-Caminho mais Barato](./assets/0005-sub-caminhos-mais-curtos.png#dark=1)
+
+### [**Relaxação**](color:yellow)
+
+Os algoritmos abordados mais abaixo irão recorrer à operação de **relaxação**. É mantido um vector, $d$, com $|V|$ elementos, onde para cada vértice $v \in V$ de $G = (V, E)$, $d[v]$ denota a estimativa atual do caminho mais curto de $s$ até $v$. Corresponderá a um **majorante** do peso do mesmo - isto é, se já sabemos que a melhor estimativa que temos é aquela, nunca iremos atualizar a mesma caso se trate de um custo mais caro. O algoritmo começa por inicializar todas as estimativas a $+\infty$, bem como todos os pais de todos os vértices a `Nil`:
+
+```rust
+InitializeSingleSource(G, s) // ocorre em \Theta(|V|) tempo
+  for each v in V[G] do
+    d[v] := +\infty
```
Comment (diogotcorreia) [inline content/asa/0005-caminhos-mais-curtos.md:107]:
```suggestion
    d[v] := +infinity
```

File: content/asa/0005-caminhos-mais-curtos.md [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+Podemos, por fim, afirmar que $\delta(s, v) \leq \delta(s, u) + w(u, v)$. Escrevemos $\leq$ em vez de $=$ porque pode haver um sub-caminho mais barato de $s$ a $v$ que não inclua necessariamente o arco $(u, v)$, como no exemplo abaixo:
+
+![Sub-Caminho mais Barato](./assets/0005-sub-caminhos-mais-curtos.png#dark=1)
+
+### [**Relaxação**](color:yellow)
+
+Os algoritmos abordados mais abaixo irão recorrer à operação de **relaxação**. É mantido um vector, $d$, com $|V|$ elementos, onde para cada vértice $v \in V$ de $G = (V, E)$, $d[v]$ denota a estimativa atual do caminho mais curto de $s$ até $v$. Corresponderá a um **majorante** do peso do mesmo - isto é, se já sabemos que a melhor estimativa que temos é aquela, nunca iremos atualizar a mesma caso se trate de um custo mais caro. O algoritmo começa por inicializar todas as estimativas a $+\infty$, bem como todos os pais de todos os vértices a `Nil`:
+
+```rust
+InitializeSingleSource(G, s) // ocorre em \Theta(|V|) tempo
```
Comment (diogotcorreia) [inline content/asa/0005-caminhos-mais-curtos.md:105]:
```suggestion
InitializeSingleSource(G, s) // ocorre em Theta(|V|) tempo
```

File: content/asa/0005-caminhos-mais-curtos.md [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+  for each v in V[G] do
+    d[v] := +\infty
+    pi[v] := Nil
+  d[s] := 0 // temos, claro, que o custo de s a s é 0
+```
+
+**Relaxar** uma aresta $(u, v)$ consiste em verificar se podemos imediatamente melhorar a nossa estimativa atual do custo de $s$ a $v$ passando pela aresta $(u, v)$ - caso possamos, atualizamos tanto a nossa estimativa como o pai de $v$:
+
+```rust
+Relax(G, u, v, w) // ocorre em \Theta(1) tempo
```
Comment (diogotcorreia) [inline content/asa/0005-caminhos-mais-curtos.md:115]:
```suggestion
Relax(G, u, v, w) // ocorre em Theta(1) tempo
```

File: content/asa/0005-caminhos-mais-curtos.md [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+
+```rust
+Dijkstra(G, w, s)
+  InitializeSingleSource(G, s) // ocorre em \Theta(V) tempo
+  let S := new array
+  let Q := new queue(V)
+  while Q is not empty do // ocorre \Theta(V) vezes
+    let u := extractMin(Q) // V vezes * O(logV) durante o algoritmo
+    S.add(u)
+    for each v in Adj[u] do // \Theta(E) vezes durante o algoritmo
```
Comment (diogotcorreia) [inline content/asa/0005-caminhos-mais-curtos.md:157]:
```suggestion
  InitializeSingleSource(G, s) // ocorre em Theta(V) tempo
  let S := new array
  let Q := new queue(V)
  while Q is not empty do // ocorre Theta(V) vezes
    let u := extractMin(Q) // V vezes * O(logV) durante o algoritmo
    S.add(u)
    for each v in Adj[u] do // Theta(E) vezes durante o algoritmo
```

**Review by diogotcorreia — CHANGES_REQUESTED** [review]

**Review by diogotcorreia — APPROVED** [review]

**Pull request merged**